### DISCUSSION TITLE
Refinements using arguments

### DISCUSSION THREAD

**Discussion by wvandermerwe** [disneystreaming/smithy4s#551] [General]
Hi there,

I have been experimenting with smithy refinements over the weekend, and was trying to get something like this to work:
```
@trait(selector: "string")
structure boundedString {
    @required
    min: Integer,
    @required
    max: Integer
}
```

Which could be applied like this:
```
@boundedString(min: 0, max: 5)
string Testy
```

The refinement type in my code would then take 2 type parameters to represent the bounds in this case. I couldn't get it to work on my side, as the type parameters were an issue when applying the refinement trait:
```
apply boundedString @refinement(
    targetType: "po.common.smithy.RefinementProviders.BoundedString",
    providerImport: "po.common.smithy.RefinementProviders.given"
)
```

I then tried to access the fields through the Shapetag in the refinement provider implicit instance, but didn't succeed with that either. I'm not sure if this is even possible at all or whether I just don't know enough about smithy(4s) to get this to work, so any help would be appreciated :)

**Comment by ahjohannessen** [comment]
@scala-impala Have you tried:

```
@length(min: 0, max: 5)
string Testy
```

**Comment by Baccata** [comment]
It is possible to access the value of the trait when constructing refinements, as such : 

```scala 
class MyConstrainedString(string: String)
def validate(min: Int, max: Int) : Surjection[String, MyConstrainedString] = ??? 


implicit val myRefinement: Refinement[BoundedString] = Refinement.drivenBy[BoundedString].contextual(boundedString => validate(boundedString.min, boundedString.max))
```

It is not possible (and **will likely never be possible**) to lift the values contained by `BoundedString` into the type system. In other words, it won't be possible for you to do `class MyConstrainedString[Min <: Int, Max <: Int](string: String)`, and for the `Min` and `Max` type parameters to be fulfilled by the min/max values from your `boundedString` trait.
